text/0000-path-clarity.md

```diff
@@ -0,0 +1,731 @@
+    will not generate a warning. This will ultimately be supplanted by [macros
+    2.0].
+  - Note that macro authors can still emit `extern crate` with an attribute to
+    silence the warning.
+
+- The compilation model for external crates is essentially the same as today:
+  they are linked only if references through an absolute path (e.g. a `use`
+  statement) or `extern crate` declaration. In either case, the fact that the
+  reference is to an external crate is completely unambiguous, and use of
+  `extern crate` is no longer needed.
+
+Note that, unlike the [previous RFC] involving `extern crate`, we do not need to
+change the interface between Cargo and `rustc` here at all, because there is no
+need to "generate" bindings based on the external crates provided. Instead, as
+today, extern crates are brought in *by demand* using absolute paths.
+
+Thus, at the end of stage 2, idiomatic code should look as follows:
+
+```rust
+use serde::Serialize;
+use crate::top_level_module::SomeItem;
+```
+
+A consequence is that you can discover all usage of external crates from a
+project's source by grepping for `use` statements or absolute paths that do not
+begin with `crate`/`super`/`self`. Put differently, usage of external crates is
+still fully explicit in the source, even without consulting `Cargo.toml`.
+
+[macros 2.0]: https://github.com/rust-lang/rfcs/blob/master/text/1561-macro-naming.md#importing-macros
+[previous RFC]: https://github.com/rust-lang/rfcs/pull/2088
+
+## Migration experience
+
+The two stages are carefully designed to retain key properties of the module
+system and compilation throughout, and to minimize the introduction of hard
+errors across epochs. The only thing that you *must* do to opt in to the next
+epoch is change absolute paths to use the new format (starts with crate name,
+`crate`, `super`, or `self`).
+
+We can and should provide a high-fidelity `rustfix` tool that performs at least
+this change, and ideally lowers `pub` to `crate` when possible as well.
+
+In the error specified for invalid fully qualified paths, in which the compiler
+checks for a top-level module with the same name, the error should *also*
+suggest that the code may be coming from an older epoch, and point to (1) the
+`rustfix` tool to automate migration and (2) the ability to specify an older
+epoch in `Cargo.toml`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The most clear drawback here is that a new epoch is required. The RFC strives to
+make the transition as narrow as possible, both by introducing as few new errors
+as possible, and by providing additional mitigations through migration tooling
+and compiler diagnostics (as well as the epoch system itself, of course).
+
+Another drawback is that imports from within your crate become more verbose,
+since they require a leading `crate`. However, this downside is considerably
+mitigated if [nesting in `use`] is permitted.
```

Comment:
... or the compiler could search both the global namespace and the crate root (for relative `use`'s only, `use ::...` would still be absolute).